src/openttd.cpp

```diff
@@ -622,19 +622,76 @@ int openttd_main(int argc, char *argv[])
 				if (mgo.opt != NULL) SetDebugString(mgo.opt);
 				break;
 			}
-		case 'e': _switch_mode = (_switch_mode == SM_LOAD_GAME || _switch_mode == SM_LOAD_SCENARIO ? SM_LOAD_SCENARIO : SM_EDITOR); break;
+		case 'e':
+			_switch_mode = (_switch_mode == SM_LOAD_GAME || _switch_mode == SM_LOAD_SCENARIO ? SM_LOAD_SCENARIO : (_switch_mode == SM_LOAD_HEIGHTMAP ? SM_LOAD_HEIGHTMAP : SM_EDITOR));
+			break;
 		case 'g':
 			if (mgo.opt != NULL) {
 				_file_to_saveload.SetName(mgo.opt);
 				bool is_scenario = _switch_mode == SM_EDITOR || _switch_mode == SM_LOAD_SCENARIO;
-				_switch_mode = is_scenario ? SM_LOAD_SCENARIO : SM_LOAD_GAME;
-				_file_to_saveload.SetMode(SLO_LOAD, is_scenario ? FT_SCENARIO : FT_SAVEGAME, DFT_GAME_FILE);
+				bool is_heightmap = _switch_mode == SM_LOAD_HEIGHTMAP;
+				_switch_mode = is_scenario ? SM_LOAD_SCENARIO : (is_heightmap ? SM_LOAD_HEIGHTMAP : SM_LOAD_GAME);
+
+				/* Unfortunately, initializing file type to being invalid breaks the SafeLoad function which expects SLO_LOAD operation: initializing to dummy value */
+				/*_file_to_saveload.SetMode(FIOS_TYPE_INVALID);
+				_file_to_saveload.SetMode(SLO_INVALID, FT_INVALID, DFT_INVALID);*/
+				_file_to_saveload.SetMode(SLO_LOAD, FT_SAVEGAME, DFT_GAME_FILE);
 
-				/* if the file doesn't exist or it is not a valid savegame, let the saveload code show an error */
+				/* If supplied file does not exist or is not a valid savegame/scenario/heightmap, let the saveload code show an error */
 				const char *t = strrchr(_file_to_saveload.name, '.');
 				if (t != NULL) {
-					FiosType ft = FiosGetSavegameListCallback(SLO_LOAD, _file_to_saveload.name, t, NULL, NULL);
-					if (ft != FIOS_TYPE_INVALID) _file_to_saveload.SetMode(ft);
+					FiosType ft;
+					ft = FiosGetSavegameListCallback(SLO_LOAD, _file_to_saveload.name, t, NULL, NULL);
+					if (ft != FIOS_TYPE_INVALID) {_file_to_saveload.SetMode(ft); _switch_mode = SM_LOAD_GAME; break;}
```

Comment:
You discard the is_scenario/is_heightmap stuff here, same for the next lines

Reply:
Those checks have been redone, along with those in blocks searching by name/title underneath. Are they better?

I seem however enable to start the editor by loading a heightmap there, irrespective of the loading method... I see a loading bar, then I get back to the game menu without any error message.
The question is: Is/Can this portion of code (be) responsible for that?